### DISCUSSION TITLE
Changing the default filter criteria to "Is equal to" - version 3.x

### DISCUSSION THREAD

**Discussion by ThisuraThejith** [sonata-project/SonataAdminBundle#8184] [Q&A]
Good day,

I'm working with Sonata Admin Panel 3.x and currently I'm working on changing the default filter matching criteria from "contains" to "is equal to" for some filter fields to filter certain entities. Refer the below screen shot to have a more clear idea.

<img width="1194" alt="Screenshot 2024-05-23 at 15 10 23" src="https://github.com/sonata-project/SonataAdminBundle/assets/15060374/6e51b905-62e7-4762-ac4a-222ae21d02d1">

I'll take one entity called "Comments" to explain my problem. The following are the filter fields that I have to filter comments.

<img width="172" alt="Screenshot 2024-05-23 at 15 12 57" src="https://github.com/sonata-project/SonataAdminBundle/assets/15060374/3d9654b0-d7bc-45d4-a010-80d00e981820">

As I need to always set the value to "**is equal to**" in the first drop down after clicking on "**Advanced filters**", I've overridden the "**configureDefaultFilterValues**" in the "**CommentAdmin**" file as follows.

```
protected function configureDefaultFilterValues(array &$filterValues)
{
      $availableFilterFields = [
          'id',
          'message',
          'user.id',
          'user.artist_name',
          'song.id',
          'song.name',
          'article.title',
      ];
    
      foreach ($availableFilterFields as $fieldName) {
          $filterValues[str_replace('.', '__', $fieldName)] = [
              "type" => ChoiceType::TYPE_EQUAL
          ];
      }
}
```

When it comes to a filter field which is a direct property of the Comment entity (for example "**message**" in this case), this works and "**is equal to**" is selected as the default filter criteria as follows.
    
<img width="1067" alt="Screenshot 2024-05-23 at 17 45 42" src="https://github.com/sonata-project/SonataAdminBundle/assets/15060374/08c3e86c-1128-439b-843e-658bbc6bf3d4">

But the problem is when it comes to a filter field which is associated to the Comment entity through another entity like User Artist Name, which will be accessed in the CommentAdmin file as **user.artist_name**, which is included in the **$availableFilterFields** array in the above method, this doesn't work and "**is equal to**" is not selected as the default filter criteria as in the following screenshot.

<img width="1067" alt="Screenshot 2024-05-23 at 17 49 05" src="https://github.com/sonata-project/SonataAdminBundle/assets/15060374/7b856057-e9b4-4f87-afa5-ff36a1fb751f">

Is there a way to resolve this? And add to that I'd appreciate if anyone can suggest a way to centrally change this default filter criteria to "**is equal to**" in every place, as that is the filter criteria that is mostly used by our system admins and it will be very convenient to have it already selected.

Thanks in advance.

**Comment by zyberspace** [comment]
Try using `user__artist_name` instead, like the actual query parameter used when submitting the filter.


The default filter values are directly merged with the passed query. There is no further processing of the provided keys, as far, as i know of.

-----

If your stil want to use the `.` in your `configureDefaultFilterValues`, you could write something like this:

```php
protected function configureDefaultFilterValues(array &$filterValues)
{
    $availableFilterFields = [
        'id',
        'message',
        'user.id',
        'user.artist_name',
        'song.id',
        'song.name',
        'article.title',
    ];

    foreach ($availableFilterFields as $fieldName) {
        $filterValues[str_replace('.', '__', $fieldName)] = [
            "type" => ChoiceType::TYPE_EQUAL
        ];
    }
}

```

**Reply by ThisuraThejith** [reply]
@zyberspace Thanks a lot for the answer and passing `user__artist_name` worked indeed. To add some more weight, is there a way that you can suggest to centrally change this default filter criteria into "is equal to" in every place in the admin panel as that is the most commonly used filter criteria by our admins?

**Reply by zyberspace** [reply]
@ThisuraThejith I would build a global admin extension: https://docs.sonata-project.org/projects/SonataAdminBundle/en/4.x/reference/extensions/

It's basically the same as your admin class but can be configured for multiple (or even all) admins and each method gets called after the one from your individual admins. The configureDefaultFilterValues method of the admin extensions gets the admin passed, so you can try and get your filter names from their and the simply put your foreach in the admin extension.
